UFW module (another take)

jkeskikangas opened the pull request:
Based on @ovcharenko #5128. Additions
- Separate `state`, `policy` and `rule` attributes
- Support for `logging` command
- Support for `direction` and `interface` attributes
- Simple and robust change notifications based on `ufw status verbose` diff

Feedback welcome.

ovcharenko committed:
New module [ufw]: this module handles Ubuntu UFW operations

* Updated documentation related to IPv6 usage.

BugFixes:
* Solved the default_policy and state mutual exclusive status.
* Fixed changed status for IPv6 addresses.

Added @otnateos patch.

Commit:
Add functionality and reduce complexity.

* Separate 'state', 'policy' and 'rule' commands
* Support for 'logging' command
* Support for 'direction' and 'interface' attributes
* Reliable change notifications based on 'ufw status verbose' diff
* Update documentation
* Cleanup

ovcharenko commented:
Good job! But I think you need create separate pull request containing ufw module only. I merged your changes... so I'm fine with whatever pull request they will select for merge.

jkeskikangas commented:
Thanks @ovcharenko! Whoa, I don't know how I messed up the pull request, it was fine a while ago. I'll see if I can fix it tomorrow. Luckily at least your's is fine.

-- edit
I fixed the pull request to contain only the relevant commits.

jkeskikangas commented:
I cleaned up the implementation and updated the docs. Ready for review.

ovcharenko commented:
Looks great!

dbrgn commented on library/system/ufw:
```diff
@@ -0,0 +1,248 @@
+      - C(enabled) reloads firewall and enables firewall on boot.
+      - C(disabled) unloads firewall and disables firewall on boot.
+      - C(reloaded) reloads firewall.
+      - C(reseted) disables and resets firewall to installation defaults.
```
Is `reseted` proper english? Shouldn't it be just `reset`? (I'm not a native speaker so I'm unsure about that :))

jkeskikangas replied:
Thanks, that's correct. I'll fix the typo.

jkeskikangas replied:
Fixed in 0707462.

dbrgn commented on library/system/ufw:
```diff
@@ -0,0 +1,248 @@
+  proto:
+    description:
+      - TCP/IP protocol.
+    choices: ['any', 'tcp', 'udp', 'ipv6']
```
What about support for `esp` and `ah` protocols (IPSec)? :) They seem to be supported according to the man page:

```
IPSec  is supported by using the 'esp' ('50') and 'ah' ('51') protocols. These protocols can only
be used with the full syntax. For example:

  ufw allow to 10.0.0.1 proto esp
  ufw allow to 10.0.0.1 from 10.4.0.0/16 proto esp
  ufw allow to 10.0.0.1 proto ah
  ufw allow to 10.0.0.1 from 10.4.0.0/16 proto ah
```

jkeskikangas replied:
Thanks, that's true. I'll add the options.

artob replied:
The list of supported protocol names surely is sourced from `/etc/protocols` and/or `/etc/services`, so adding support for protocol names one by one might not be a tenable approach?

jkeskikangas replied:
Thanks @dbrgn, support for `esp` and `ah` added in ca2ab4c. 

@bendiken iptables (and ufw) works on transport layer (tcp, udp) and Internet layer (ip, ipsec) protocols. /etc/services actually maps well-known service names to tcp/udp ports, as explained in the [man page](http://manpages.ubuntu.com/manpages/precise/man5/services.5.html).

dbrgn commented:
Something I noticed: When deploying some rules onto a new server, ansible did not report the tasks as "changed" even though the firewall rules were newly created (and work). Might that be due to the firewall being disabled when adding the rules?

Otherwise the module seems to work nicely. Thanks for creating it.

Commit:
Add protocol options 'esp' and 'ah'.

Commit:
Fix typo: 'reseted' -> 'reset'.

anderiv commented:
This looks great! Has it been targeted to a release yet?

I was hoping it would make it into 1.5, but doesn't look like that happened.

Commit:
Insert rule to a specific num.

jkeskikangas commented:
@dbrgn, the ansible 'changed' notifications for the ufw module are based on diffing the output of `ufw status verbose` before and after running the actual command.

If the firewall is disabled as new rules are added, they are not regarded as changing the (running) state of the firewall. Please see the following outputs run on two clean slate ubuntu 12.04:

**Firewall enabled as a first step**

```
TASK: [base | Enable firewall] ************************************************
changed: [foo]

TASK: [base | Allow tcp/22 for SSH] *******************************************
changed: [foo]

TASK: [base | Reject other ports] *********************************************
changed: [foo]
```

**Firewall enabled as a last step**

```
TASK: [base | Allow tcp/22 for SSH] *******************************************
ok: [bar]

TASK: [base | Reject other ports] *********************************************
ok: [bar]

TASK: [base | Enable firewall] ************************************************
changed: [bar]
```

dbrgn commented:
Can we maybe fix that behavior? I think a change should be reflected (as the rules _are_ changed, they're just not enabled.)

jkeskikangas commented:
It seems that diffing `/lib/ufw/user*.rules` in addition to `ufw status` might work: http://askubuntu.com/questions/30781/see-configured-rules-even-when-inactive

I'll add an additional check by running 

```
grep '^### tuple' /lib/ufw/user*.rules
```

before and after executing the ufw command.

Commit:
Recognize rule changes even if ufw is in disabled state.

See http://askubuntu.com/questions/30781/see-configured-rules-even-when-inactive
for the details.

jkeskikangas commented:
@dbrgn rule changes are now recognized properly even if ufw is in disabled state. Implemented in 37ef39d.

jimi-c commented:
Merged, thanks!
